src/stash/src/upgrade/v33_to_v34.rs

```diff
@@ -94,124 +100,140 @@ async fn migrate_cluster_replicas(tx: &mut Transaction<'_>) -> Result<(), StashE
+    let clusters = tx
+        .peek_one::<v33::ClusterKey, v33::ClusterValue>(
+            tx.collection(CLUSTER_COLLECTION.name()).await?,
+        )
+        .await?;
+
+    // We need to allocate additional audit events, which need IDs. Gather the current next_id.
     let coll = ID_ALLOCATOR_COLLECTION.from_tx(tx).await?;
     let key = v33::IdAllocKey {
         name: AUDIT_LOG_ID_ALLOC_KEY.to_string(),
     };
     let mut next_audit_log_id = tx.peek_key_one(coll, &key).await?.unwrap().next_id;
 
-    // Collect all existing audit log events.
-    let coll = AUDIT_LOG_COLLECTION.from_tx(tx).await?;
-    let audit_events = tx.peek_one(coll).await?;
-    let mut events: Vec<_> = audit_events
-        .into_keys()
-        .map(|key| {
-            let v33::audit_log_key::Event::V1(event) = key.event.unwrap();
-            event
-        })
-        .collect();
-
-    // Find live replicas by iterating through the events in order and keeping track of the
-    // set of replicas not yet dropped. When every event is processed, this set contains
-    // only the currently live replicas.
-    events.sort_by_key(|e| e.id);
-    let mut live_replicas = BTreeMap::new();
-    for event in events {
-        let details = event.details.unwrap();
-        match details {
-            v33::audit_log_event_v1::Details::CreateClusterReplicaV1(details) => {
-                let key = (details.cluster_id.clone(), details.replica_id.clone());
-                let prev = live_replicas.insert(key, details);
-                assert_eq!(prev, None, "replica created twice");
-            }
-            v33::audit_log_event_v1::Details::DropClusterReplicaV1(details) => {
-                let key = (details.cluster_id, details.replica_id);
-                let prev = live_replicas.remove(&key);
-                assert!(prev.is_some(), "non-existing replica dropped");
-            }
-            _ => (),
+    let mut new_events = Vec::with_capacity(entries.len());
+    for (old_id, new_key, new_value) in entries {
+        let cluster_id = match new_value.cluster_id {
+            Some(v34::ClusterId {
+                value: Some(v34::cluster_id::Value::User(id)),
+            }) => format!("u{id}"),
+            Some(v34::ClusterId {
+                value: Some(v34::cluster_id::Value::System(id)),
+            }) => format!("s{id}"),
+            Some(v34::ClusterId { value: None }) | None => "".to_string(),
         };
-    }
-
-    // For each live replica, add a `drop` event with the old ID and a `create` event with
-    // the new ID.
-    let mut updates = Vec::with_capacity(live_replicas.len() * 2);
-    for (_, details) in live_replicas {
-        let old_id = details.replica_id.unwrap().inner;
-        let numeric_id = old_id.parse::<u64>().unwrap();
-        assert!(
-            details.cluster_id.starts_with('u'),
-            "non-user cluster in audit_log"
-        );
```

Comment:
@teskje is there a reason we asserted this? I'm pretty sure we would have failed this assertion on at least one user's environment

Reply:
It was there because I assumed we don't log system replicas to the audit log, because we don't want to bill for them, so I figured we can avoid handling this case. As you found out my assumption was wrong and there are some envs with system replicas in the audit log. I assume these come from us manually creating and dropping system replicas in user envs.